File: backend/src/test/java/com/kiwi/combat/CombatTestFactory.java
package com.kiwi.combat;

import com.kiwi.features.combat.data.dto.CombatDTO;
import com.kiwi.features.combat.data.dto.CombatTurnResultDTO;
import com.kiwi.features.combat.data.enums.BarkDismissMode;
import com.kiwi.features.combat.data.enums.BarkTriggerType;
import com.kiwi.features.combat.data.enums.CombatActorType;
import com.kiwi.features.combat.data.enums.CombatActionType;
import com.kiwi.features.combat.data.enums.CombatGeneralStatus;
import com.kiwi.features.combat.data.persistence.*;

import java.time.Instant;
import java.util.List;
import java.util.Random;

public class CombatTestFactory {

    private static final Random RANDOM = new Random();

    private static int randStat() {
        return 50 + RANDOM.nextInt(100);
    }

    private static int randHp() {
        return 200 + RANDOM.nextInt(800);
    }

    // =========================================================================
    // USER STATS
    // =========================================================================

    public static UserStatsPersistence userStats(Long userId) {
        return UserStatsPersistence.builder()
                .userId(userId)
                .maxHp(randHp())
                .shield(randStat())
                .patk(randStat())
                .matk(randStat())
                .pdef(randStat())
                .mdef(randStat())
                .acc(randStat())
                .eva(randStat())
                .lck(randStat())
                .build();
    }

    // =========================================================================
    // ENEMY
    // =========================================================================

    public static EnemyPersistence enemy(Long id) {
        return EnemyPersistence.builder()
                .id(id)
                .name("Enemy " + id)
                .sprite("sprite_" + id)
                .maxHp(randHp())
                .shield(randStat())
                .patk(randStat())
                .matk(randStat())
                .pdef(randStat())
                .mdef(randStat())
                .acc(randStat())
                .eva(randStat())
                .lck(randStat())
                .build();
    }

    // =========================================================================
    // COMBAT CONFIG
    // =========================================================================

    public static CombatConfigPersistence combatConfig(Long id, Long enemyId) {
        return CombatConfigPersistence.builder()
                .id(id)
                .enemyId(enemyId)
                .timeLimit(5 + RANDOM.nextInt(30))
                .build();
    }

    // =========================================================================
    // COMBAT
    // =========================================================================

    public static CombatPersistence ongoingCombat(
            Long userId,
            Long configId,
            UserStatsPersistence stats,
            EnemyPersistence enemy
    ) {
        return combat(userId, configId, stats, enemy, CombatGeneralStatus.ONGOING);
    }

    public static CombatPersistence finishedCombat(
            Long userId,
            Long configId,
            UserStatsPersistence stats,
            EnemyPersistence enemy
    ) {
        return combat(userId, configId, stats, enemy, CombatGeneralStatus.USER_WON);
    }

    private static CombatPersistence combat(
            Long userId,
            Long configId,
            UserStatsPersistence stats,
            EnemyPersistence enemy,
            CombatGeneralStatus status
    ) {
        return CombatPersistence.builder()
                .userId(userId)
                .combatConfigId(configId)
                .enemyId(enemy.getId())

                .userHp(stats.getMaxHp())
                .userMaxHp(stats.getMaxHp())
                .userShield(stats.getShield())
                .userPatk(stats.getPatk())
                .userMatk(stats.getMatk())
                .userPdef(stats.getPdef())
                .userMdef(stats.getMdef())
                .userAcc(stats.getAcc())
                .userEva(stats.getEva())
                .userLck(stats.getLck())
                .userTurns(0)

                .enemyHp(enemy.getMaxHp())
                .enemyMaxHp(enemy.getMaxHp())
                .enemyShield(enemy.getShield())
                .enemyPatk(enemy.getPatk())
                .enemyMatk(enemy.getMatk())
                .enemyPdef(enemy.getPdef())
                .enemyMdef(enemy.getMdef())
                .enemyAcc(enemy.getAcc())
                .enemyEva(enemy.getEva())
                .enemyLck(enemy.getLck())
                .enemyTurns(0)

                .turnNumber(1)
                .endsAt(Instant.now().plusSeconds(600))
                .combatStatus(status)
                .build();
    }

    // =========================================================================
    // COMBAT LOG
    // =========================================================================

    public static CombatLogPersistence combatLog(Long combatId, int turnNumber, CombatActorType actor) {
        return CombatLogPersistence.builder()
                .combatId(combatId)
                .turnNumber(turnNumber)
                .actor(actor)
                .combatActionType(CombatActionType.SKILL_USED)
                .skillName("Skill " + RANDOM.nextInt(1000))
                .createdAt(Instant.now())
                .build();
    }

    // =========================================================================
    // LAST SKILL
    // =========================================================================

    public static CombatLastSkillPersistence lastSkill(Long combatId, CombatActorType actor, Long skillId) {
        return CombatLastSkillPersistence.builder()
                .id(new CombatLastSkillKey(combatId, actor))
                .skillId(skillId)
                .build();
    }

    // =========================================================================
    // BLOCKED SKILL
    // =========================================================================

    public static CombatBlockedSkillPersistence blockedSkill(Long combatId, CombatActorType actor, Long skillId) {
        return CombatBlockedSkillPersistence.builder()
                .id(new CombatBlockedSkillKey(combatId, actor, skillId))
                .build();
    }

    // =========================================================================
    // BARK TRIGGER
    // =========================================================================

    public static CombatBarkTriggerPersistence barkTrigger(
            Long combatConfigId,
            BarkTriggerType type,
            Long conversationId
    ) {
        return CombatBarkTriggerPersistence.builder()
                .combatConfigId(combatConfigId)
                .type(type)
                .threshold(type == BarkTriggerType.SKILL_USED ? null : 50f)
                .skillId(type == BarkTriggerType.SKILL_USED ? 1L : null)
                .conversationId(conversationId)
                .dismissMode(BarkDismissMode.AUTO)
                .priority(0)
                .build();
    }

    public static CombatFiredBarkPersistence firedBark(Long combatId, Long triggerId) {
        return CombatFiredBarkPersistence.builder()
                .id(new CombatFiredBarkKey(combatId, triggerId))
                .build();
    }

    // =========================================================================
    // DTOs
    // =========================================================================

    public static CombatDTO combatDTO(Long id) {
        return CombatDTO.builder()
                .id(id)
                .combatConfigId(1L + RANDOM.nextInt(100))
                .turnNumber(1 + RANDOM.nextInt(10))
                .combatStatus(CombatGeneralStatus.ONGOING.name())
                .endsAt(Instant.now().plusSeconds(600).toEpochMilli())
                .enemyName("Enemy " + id)
                .enemySprite("sprite_" + id)
                .log(List.of())
                .build();
    }

    public static CombatTurnResultDTO combatTurnResultDTO(Long combatId) {
        return CombatTurnResultDTO.builder()
                .combatId(combatId)
                .turnNumber(1 + RANDOM.nextInt(10))
                .actions(List.of())
                .combatStatus(CombatGeneralStatus.ONGOING.name())
                .createdAt(Instant.now())
                .build();
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/combat/tests/CombatTestFactory.kt
package com.bellako.kiwi.features.combat.tests

import com.bellako.kiwi.features.combat.data.CombatActionDomain
import com.bellako.kiwi.features.combat.data.CombatActionType
import com.bellako.kiwi.features.combat.data.CombatActor
import com.bellako.kiwi.features.combat.data.CombatActorDomain
import com.bellako.kiwi.features.combat.data.CombatDomain
import com.bellako.kiwi.features.combat.data.CombatGeneralStatus
import com.bellako.kiwi.features.combat.data.CombatStatsDomain
import com.bellako.kiwi.features.combat.data.CombatTurnResultDomain
import com.bellako.kiwi.features.combat.data.SkillEffectResultDomain
import com.bellako.kiwi.features.combat.data.SkillEffectResultType

@Suppress("MagicNumber", "LongParameterList", "TooManyFunctions")
object CombatTestFactory {
    fun validCombatStatsDomain(
        currentHp: Int = 100,
        maxHp: Int = 100,
        patk: Int = 20,
        matk: Int = 20,
        pdef: Int = 10,
        mdef: Int = 10,
        acc: Int = 90,
        eva: Int = 10,
        lck: Int = 5,
    ): CombatStatsDomain =
        CombatStatsDomain(
            currentHp = currentHp,
            maxHp = maxHp,
            patk = patk,
            matk = matk,
            pdef = pdef,
            mdef = mdef,
            acc = acc,
            eva = eva,
            lck = lck,
        )

    fun validCombatActorDomain(stats: CombatStatsDomain = validCombatStatsDomain()): CombatActorDomain =
        CombatActorDomain(
            stats = stats,
            elementalMultipliers = emptyList(),
            statusResistances = emptyList(),
            activeStatus = emptyList(),
        )

    fun validCombatDomain(
        id: Long = 1L,
        combatConfigId: Long = 1L,
        turnNumber: Int = 0,
        endsAt: Long? = null,
        combatStatus: CombatGeneralStatus = CombatGeneralStatus.ONGOING,
        enemyName: String = "Test Enemy",
        enemySprite: String = "enemy_sprite",
        backgroundId: Long? = null,
        user: CombatActorDomain = validCombatActorDomain(),
        enemy: CombatActorDomain = validCombatActorDomain(),
        log: List<CombatActionDomain> = emptyList(),
        onCompletedEvent: String? = null,
        onCompletedEntityId: Int? = null,
    ): CombatDomain =
        CombatDomain(
            id = id,
            combatConfigId = combatConfigId,
            turnNumber = turnNumber,
            endsAt = endsAt,
            combatStatus = combatStatus,
            enemyName = enemyName,
            enemySprite = enemySprite,
            backgroundId = backgroundId,
            user = user,
            enemy = enemy,
            log = log,
            onCompletedEvent = onCompletedEvent,
            onCompletedEntityId = onCompletedEntityId,
        )

    fun skillUsedAction(
        actor: CombatActor = CombatActor.USER,
        skillName: String = "Fireball",
        damage: Float = 25f,
    ): CombatActionDomain =
        CombatActionDomain(
            actor = actor,
            actionType = CombatActionType.SKILL_USED,
            skillName = skillName,
            skillEffectsResults =
                listOf(
                    SkillEffectResultDomain(
                        typeResult = SkillEffectResultType.DAMAGE,
                        target = if (actor == CombatActor.USER) CombatActor.ENEMY else CombatActor.USER,
                        value = damage,
                        critic = false,
                    ),
                ),
        )

    fun timeoutAction(): CombatActionDomain =
        CombatActionDomain(
            actor = CombatActor.USER,
            actionType = CombatActionType.TIMEOUT,
        )

    fun abandonAction(): CombatActionDomain =
        CombatActionDomain(
            actor = CombatActor.USER,
            actionType = CombatActionType.ABANDON,
        )

    fun validCombatTurnResultOngoing(
        combatId: Long = 1L,
        turnNumber: Int = 1,
        actions: List<CombatActionDomain> = listOf(skillUsedAction()),
    ): CombatTurnResultDomain =
        CombatTurnResultDomain(
            combatId = combatId,
            turnNumber = turnNumber,
            actions = actions,
            combatStatus = CombatGeneralStatus.ONGOING,
            onCompletedEvent = null,
            onCompletedEntityId = null,
        )

    fun userWonTurnResult(
        combatId: Long = 1L,
        turnNumber: Int = 5,
        onCompletedEvent: String? = "COMPLETE_QUEST",
        onCompletedEntityId: Int? = 42,
        actions: List<CombatActionDomain> = listOf(skillUsedAction(damage = 100f)),
    ): CombatTurnResultDomain =
        CombatTurnResultDomain(
            combatId = combatId,
            turnNumber = turnNumber,
            actions = actions,
            combatStatus = CombatGeneralStatus.USER_WON,
            onCompletedEvent = onCompletedEvent,
            onCompletedEntityId = onCompletedEntityId,
        )

    fun userLostTurnResult(
        combatId: Long = 1L,
        turnNumber: Int = 5,
        onCompletedEvent: String? = "COMPLETE_QUEST",
        onCompletedEntityId: Int? = 42,
        actions: List<CombatActionDomain> = listOf(skillUsedAction(actor = CombatActor.ENEMY, damage = 200f)),
    ): CombatTurnResultDomain =
        CombatTurnResultDomain(
            combatId = combatId,
            turnNumber = turnNumber,
            actions = actions,
            combatStatus = CombatGeneralStatus.USER_LOST,
            onCompletedEvent = onCompletedEvent,
            onCompletedEntityId = onCompletedEntityId,
        )

    fun timeoutTurnResult(
        combatId: Long = 1L,
        turnNumber: Int = 3,
        onCompletedEvent: String? = "COMPLETE_QUEST",
        onCompletedEntityId: Int? = 42,
    ): CombatTurnResultDomain =
        CombatTurnResultDomain(
            combatId = combatId,
            turnNumber = turnNumber,
            actions = listOf(timeoutAction()),
            combatStatus = CombatGeneralStatus.USER_LOST,
            onCompletedEvent = onCompletedEvent,
            onCompletedEntityId = onCompletedEntityId,
        )

    fun abandonTurnResult(
        combatId: Long = 1L,
        turnNumber: Int = 3,
        onCompletedEvent: String? = "COMPLETE_QUEST",
        onCompletedEntityId: Int? = 42,
    ): CombatTurnResultDomain =
        CombatTurnResultDomain(
            combatId = combatId,
            turnNumber = turnNumber,
            actions = listOf(abandonAction()),
            combatStatus = CombatGeneralStatus.USER_LOST,
            onCompletedEvent = onCompletedEvent,
            onCompletedEntityId = onCompletedEntityId,
        )
}
